victor-ferro/instantclip
# Feature auth updates

### :tophat: What is the goal?
* Improve authentication
### :memo: How is it being implemented?
* encrypting passwords in db
* hiding jwt secret key as environment variable
* add provider to implement bcrypt functions

### Usage
```typescript
@Injectable()
export class BcryptOperator{
    constructor(){}

    async createHash(pass: string){
        return await bcrypt.hash(pass, await bcrypt.genSalt());
    }
    
    async compare(pass: string, hash: string){
        return await bcrypt.compare(pass, hash);
    }
}
```

```typescript
@Injectable()
export class AuthService {
    constructor(
        private usersService: UsersService,
        private jwtService: JwtService
        ){}

    async validateUser(username: string, pass: string): Promise<any> {
        const user = await this.usersService.findOne(username)
        if(user && (await this.usersService.compare(pass, user.password))){
            const { password, ...result } = user;
            return result;
        }
        return null;
    }
```


### :boom: How can it be tested?
* Run ser server (npm run start)
* make post to create user. --> curl -X POST http://localhost:3000/users -d '{"name": "lebron", "username":"kingJames","avatarUrl": "clavelandJames", "password": "cleveland"}' -H "Content-Type: application/json"
* You could see the user in db with his hashpassword and also in the console where you run the server
